keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +337,302 @@ or later. The dual-stack feature requires no changes to this API.
+In a dualstack cluster Services can be:
+1. Single Stack: service ip are allocated (or reserved, if provided by user) from the first or the second cidr (if configured) entry as configured using `--service-cluster-ip-range` flag, in this scenario services will have a single allocated `ClusterIP` (also set in `ClusterIPs`).
+2. Dual Stack (optional): service ip is allocated (or reserved, if provided by user) from both the first and the second cidr (if configured) entries. In this scenario services will have a dual assigned `ClusterIPs`. If the cluster is not configured for dualstack then  the resulting service will be created as a single stack and `ClusterIP/ClusterIPs` will be assigned from a single family.
+3. Dual Stack (required): if the cluster is not configured for dual stack then service creation will fail. If the cluster is configured for dual stack the resulting service will carry two dual stack `ClusterIPs`.
+
+The above is achieved using the following changes to Service api.
+1. Add a `spec.preferDualStack` optional boolean field.
+2. Add a `spec.ipFamilies[]` optional field with possible values of `IPv4`, `IPv6` for each each entry or `nil` for the entire slice. Maximum entries in this array is two and each entry must be `IPv4`or `IPv6` in any order. This field identifies the expected ip families of a service. api-server will reject the service creation request if the ip family is not supported by the cluster. Further dependency and validation on this field is described below.
+3. Add a `spec.ClusterIPs[]` field that follows the same rules as `status.PodIPs/status.PodIP`. Maximum entries in this array is two entries of two different ip address families. Semantically `spec.ClusterIP` will be referred as `Primary Service IP`.
+
+> Services are expected to be semantically valid whenever values for the optional fields are provided. An example a service with `spec.ipFamilies : [ ipv4 ]` and `spec.CluterIPs: [ 'some-ipv6' ]` is considered invalid because the family does not match the `ClusterIPs[0]`.
 
-For example, the following Service specification includes the `spec.ipFamily` field. Kubernetes will assign an IPv6 address (also known as a “cluster IP”) from the configured service-cluster-ip-range to this Service.
+
+The rules that govern these fields are described in the below sections
+
+##### Creating a New Single Stack Service
+
+User who want to create a new single stack Service can create it using one of the following methods:
+1. Set `spec.preferDualStack: nil` and `spec.ipFamilies: nil` fields. api-server will default `spec.preferDualStack: false` and `spec.ipFamilies: ['default ip family of the cluster']`.
+2. Set `spec.preferDualStack: false` api-server will default `spec.ipFamilies: ['cluster default ip address family']`.
```

Comment:
Do we really need to distinguish `nil` from `false` for `preferDualStack`?

In particular, the distinction is _not_ needed for the mutating admission hook case like it was with `ipFamilyPolicy`, because if you want to say "I can't be mutated to dual-stack" you do that by setting `ipFamilies`, not by setting `preferDualStack` (because explicitly setting `preferDualStack: false` but not setting `ipFamilies` doesn't make sense)

If we do need both `nil` and `false` then the two examples should be consistent. (specifically "default ip family of the cluster" vs "cluster default ip address family")

Reply:
I don't think we need to handle nil.  We can safely use defaulting to convert nil to false (or make it a non-pointer bool)

Reply:
only at ipalloc, not in the defaulting logic. To allow web hooks to change `single stack by default` to `dual stack by default`

Reply:
I'm not sure exactly which part of which comment you're referring to, but what I was saying is that we _do not_ need to have a `nil` vs `false` distinction, at any point, including to handle the webhook case.

`preferDualStack: false` means what `ipFamilyPolicy: nil` used to, which means that a webhook can change a `preferDualStack: false` Service to `preferDualStack: true` as long as `ipFamilies` or `clusterIPs` isn't set, regardless of whether it's `preferDualStack: false` because the user specified that explicitly or because the value was filled in by default.

Reply:
The basic idea is to allow webhooks to identify 
1. `preferDualStack:nil + nil families || single family entry => no preference. can do both. i don't have code that depend on single/multiple IPs`
2. `preferDualStack:false + nil families || single family entry => I can not work with dual stack, i will fail if it is dual stack`
3. `preferDualStack:true + nil families || single family => i can work with dual stack if available`
4.  `preferDualStack:true + two ip families => i have to have dual stack`

a web hook can not and *must* not change `false` => `true` for 
1. This is *not* caller intention.
2. It can not tell if the caller can actually deal with dual stack or not.

> forcing `false` => `true` might be applicable to some use cases, but it won't be the default use case.

Reply:
> `preferDualStack:false && (nil families || single family entry) => I can not work with dual stack, i will fail if it is dual stack`

No. This was the point of getting rid of `ipFamilyPolicy: SingleStack` and switching to a two-valued field (`false`/`true`) rather than a three-valued field (`nil`/`SingleStack`/`DualStack`). There is no "I can't do dual stack" any more. You have to say either "I can only do single-stack IPv4" (`preferDualStack: false, ipFamilies: ["IPv4"]`) or "I can only do single-stack IPv6" (`preferDualStack: false, ipFamilies: ["IPv6"]`). It doesn't make sense to say "I can't do dual-stack but I have no preference about IP families". Thus, there's no reason to treat `preferDualStack: nil, ipFamilies: nil` and `preferDualStack: false, ipFamilies: false` as meaning different things, and thus there is no reason to even have `preferDualStack: nil`.

Reply:
Or whatever. We're still alpha. We can still fix this in 1.20 if it merges one way now and we agree later it was wrong. We kind of need to get this merged soon though.

Reply:
Reviewing now, but the problem is defaulting.

We want to be able to asset that consumers only need to look at `ipFamilies` and that they can rely on `ipFamilies` to not be empty.

That means that an old, stored service MUST be defaulted properly on read.

For a headful service, I can rely on `clusterIP`, but for a headless Service I can't.

So defaulting code observes `{ clusterIP: None, ipFamilies: nil }` it has to convert that to `{ clusterIP: None, ipFamiles: [ $default-family] }`.  Defaulting can't tell the difference between reading an old saved instance or a new create op.  Admission control runs AFTER defaulting, so this hypothetical admission controller can not know whether the user meant to say "I require single-stack v4" or "I don't care, but the defaults have been set".

We either need a tri-state (blech) or we need to distinguish read from create when applying defaults. 

Kal was looking at this last option last I spoke to him.

Reply:
If we can set ipFamilies in code understands get vs create, does the nil matter for prefer?  What I mean is this:
* on create, user says (<nothing>, nil, nil); defaulting doesn't touch it; admission controller sees (false, nil, nil); allocator sets (false, [$default], [IP])
* on reading an old value we start with (<nothing>, nil, nil); defaulting doesn't touch it; REST.Get() sets (false, [$default], [IP])

So reading an old, saved instance gets defaulted correctly in the read path, and a new instance gets assigned correctly on the write path, but the admission controller knows that (false, nil, nil) is unambiguously "available to be changed"

The main point being that point being that pointer-ness of the bool doesn't matter FOR THIS. We may still make it a pointer because convention, but then defaulting can simply turn nil -> false (there's an ongoing debate about Go zero-values vs optional fields)